CDI Lite packaging, deployment and lifecycle specification (#546)
jakartaee/cdi: spec/src/main/asciidoc/core/spi_lite.asciidoc

[[spi_lite]]

== Build compatible extensions

A build compatible extension may integrate with the container during deployment time, as defined in <<initialization>>.

> Comprehensive specification of build compatible extensions is available in the javadoc.

Reviewer: No other part of spec refers to javadoc, I wouldn't do it here either. The overview you wrote is IMO informative enough.

Author: There's actually _one_ other part of the spec that does refer to a javadoc :-) But if the javadoc text is implicitly assumed to be part of the spec, then sure.

This text provides an overview.

[[bce]]

=== The `BuildCompatibleExtension` interface

A build compatible extension is a service provider of the `jakarta.enterprise.inject.build.compatible.spi.BuildCompatibleExtension` interface, declared in `META-INF/services`.

[source, java]
----
public interface BuildCompatibleExtension {}
----

Build compatible extensions can define arbitrary `public`, non-`static`, `void`-returning methods without type parameters, annotated with one of the extension annotations.
Such methods are called _extension methods_.

> Extension annotations correspond to extension execution phases:

Reviewer: ```suggestion
Extension method annotations correspond to extension execution phases:
```

Author: The previous paragraph introduces the term "extension annotation". I'm not entirely happy about introducing so many terms, but I can't think of a better way.

Author: Not done, to avoid breaking the flow of the text.

* `@Discovery`
* `@Enhancement`
* `@Registration`
* `@Synthesis`
* `@Validation`

Extension methods can declare arbitrary number of parameters.

> Which parameters can be declared depends on the particular execution phase.

Reviewer: ```suggestion
Each execution phase dictates what parameters can be declared.
```

Author: Going with "In each execution phase, different types of parameters may be declared."

> All the parameters will be provided by the container when the extension method is invoked.

Reviewer: Maybe mention what happens if you try to declare a parameter that isn't allowed?

> The container instantiates a single instance of each build compatible extension and invokes all extension methods on this instance.

Reviewer: ```suggestion
The container instantiates a single instance of each build compatible extension, ensuring that all extension methods are invoked on the same instance.
```

Reviewer: Message is the same. I only suggested a change because your version was harder to read for me :D

> The invocation order for extension methods may be declared using the `@Priority` annotation.

Reviewer: ```suggestion
The invocation order for extension methods may be controlled using the `@Priority` annotation.
```

Author: Will do.

> If an extension method does not have the `@Priority` annotation, the default priority `jakarta.interceptor.Interceptor.Priority.APPLICATION + 500` is assumed.

Reviewer: Add a note that if two methods have equal priority, the ordering is undefined?

Author: It's in the javadoc, but I'll add one here too.

Extension methods annotated `@Discovery` may declare parameters of the following types:

* `ScannedClasses`
* `MetaAnnotations`
* `Messages`

[source, java]
----
public interface ScannedClasses {
    void add(String className);
}
----

[source, java]
----
public interface MetaAnnotations {
    ClassConfig addQualifier(Class<? extends Annotation> annotation);
    ClassConfig addInterceptorBinding(Class<? extends Annotation> annotation);
    ClassConfig addStereotype(Class<? extends Annotation> annotation);

    void addContext(Class<? extends Annotation> scopeAnnotation, Class<? extends AlterableContext> contextClass);
    void addContext(Class<? extends Annotation> scopeAnnotation, boolean isNormal, Class<? extends AlterableContext> contextClass);
}
----

If the `addQualifier`, `addInterceptorBinding` or `addStereotype` method is called, the return value allows configuring meta-annotations on the qualifier, interceptor binding or stereotype annotation and its members.

[source, java]
----
public interface Messages {
    void info(String message);
    void info(String message, AnnotationTarget relatedTo);
    void info(String message, BeanInfo relatedTo);

>     void info(String message, ObserverInfo relatedTo);

Reviewer: Hmm, you cannot get hold of `BeanInfo` at this point (and `ObserverInfo`). I understand you are using the same class for multiple phases but this is rather confusing. I'd probably just say that `Messages` class should be used for logging or throwing an error and leave it at that without showing its code.
Or perhaps show it separately at a later stage once you introduce all  phases?

    MethodConfig addAnnotation(Annotation annotation);
    MethodConfig removeAnnotation(Predicate<AnnotationInfo> predicate);
    MethodConfig removeAllAnnotations();

    List<ParameterConfig> parameters();
}

public interface ParameterConfig extends DeclarationConfig {
    ParameterInfo info();

    ParameterConfig addAnnotation(Class<? extends Annotation> annotationType);
    ParameterConfig addAnnotation(AnnotationInfo annotation);
    ParameterConfig addAnnotation(Annotation annotation);
    ParameterConfig removeAnnotation(Predicate<AnnotationInfo> predicate);
    ParameterConfig removeAllAnnotations();
}

public interface MethodInfo extends DeclarationInfo {
    String name();
    List<ParameterInfo> parameters();
    Type returnType();
    Type receiverType();
    List<Type> throwsTypes();
    List<TypeVariable> typeParameters();

    boolean isConstructor();
    boolean isStatic();
    boolean isAbstract();
    boolean isFinal();
    int modifiers();

    ClassInfo declaringClass();
}

public interface ParameterInfo extends DeclarationInfo {
    String name();
    Type type();

    MethodInfo declaringMethod();
}
----

> When the extension method declares a parameter of type `MethodConfig` or `MethodInfo`, it will be called for each method and constructor of each discovered class matching the criteria defined on the `@Enhancement` annotation.

Reviewer: `When the extension method` -> `When an extension method`?
Not sure but `the` doesn't sound right since you are not referring to one specific method. If you choose to change it, note that it is repeated several times in this file.
